immunopepper/dna_to_peptide.pyx

```diff
@@ -34,25 +27,23 @@ def translate_dna_to_peptide(dna_str:str,all_read_frames:bool):
         'TGC': 'C', 'TGT': 'C', 'TGA': '_', 'TGG': 'W'
     }
     dna_str = dna_str.upper()
-    cdef list multiple_pep = []
+    cdef list peptides = []
     cdef bint has_stop_codon = False
     cdef list aa_str = []
-    for idx in range(0, len(dna_str), 3):
+    for idx in range(0, (len(dna_str) // 3) * 3, 3):
         codon = dna_str[idx:idx + 3]
-        if len(codon) < 3:
-            break
-        if 'N' in codon:
+        if codon[0] == 'N' or codon[1] == 'N' or codon[2] == 'N':
```

Comment:
Typical `if 'N' in codon` gets translated into a for loop across the string, which requires an additional index, and a bounds check at each iteration, so unrolling the loop is typically faster. I haven't really checked how smart the compiler is here, I just assumed it wasn't smart enough to unroll the loop.

Reply:
Thanks for the explanation. At least my quick empirical test resulted in an advatage for the `'N' in codon` construct:

```
codon = 'ACG'
 %timeit -n 10000 (codon[0] == 'N' or codon[1] == 'N' or codon[2] == 'N')
220 ns ± 5.73 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
%timeit -n 10000 'N' in codon
41.1 ns ± 0.197 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

codon = 'NNN'
%timeit -n 10000 (codon[0] == 'N' or codon[1] == 'N' or codon[2] == 'N')
62.4 ns ± 4.26 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
In [12]: %timeit -n 10000 'N' in codon
43 ns ± 0.455 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

Reply:
Nice! Note that this is pure python, while the code here is cpython (gets translated into C), so the performance gain may not reflect reality. I'll check the generated c code to see which one is faster.

Reply:
Ah, good point. I hadn't caught the `*.pyx`. In C your version might indeed be faster.

Reply:
Even in CPython the original version is 0.1-0.3% faster, so I'll leave it as is. The reason is that the C translation of CPython turns the comparison of 'N' with codon[0] into a string comparison instead of simply comparing 2 bytes.

Do you have a sense how much of a bottleneck this method is? I can just re-write it completely in C and it'll probably be significantly faster, but it's only worth the effort (will take me about 1h) if it will make an overall difference.

Reply:
I don't think this is the bottleneck. So not worth the effort.

Reply:
I did run profiling some time ago, and a lot of time was spent inside this function. Hence the cpython code. However, I think that now the bottleneck is probably somewhere below the _for_ loop of _traversal:get_and_write_peptide_and_kmer_ when the number of exon pairs to loop through gets huge